### PR TITLE
Integrate LoRA into the codebase

### PR DESCRIPTION
**Pull request by TearGosling**
LoRA promises significant reductions in memory. Hopefully, it will help take us to higher-scale models such as 12B and 20B. Right now, this is fine as is, but with the one caveat that I'm not exactly sure how to have it such that the LoRA weights are properly merged with the model. `merge_weights=True` in the `LoraConfig` doesn't do the trick, it seems. I'll be looking into it further, but in the meantime, this should be good to at least test out LoRA's effects on VRAM usage.

### CODE REVIEW

File: src/training/sft.py [hunk 1]
```diff
@@ -158,7 +161,39 @@ def save_model(self) -> None:
...
+            low=1,
+            high=model_config.vocab_size + 1,
+            # max_position_embeddings = 2048
+            size=(self.args.batch_size, model_config.max_position_embeddings)
+        ).to("cuda")
+        attention_mask = torch.zeros((self.args.batch_size, model_config.max_position_embeddings)).to("cuda")
+        # Start positions can be 0 and end positions can be 1, it doesn't really matter.
+        start_positions = torch.zeros((self.args.batch_size, 1)).to("cuda")
+        end_positions = torch.ones((self.args.batch_size, 1)).to("cuda")
+        rewards = torch.ones((self.args.batch_size, 1)).to("cuda")
```
Comment (0x000011b) [inline src/training/sft.py:177]: Not important for this case since it's a single warmup step, but just for future reference:

The [PyTorch documentation](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#create-tensors-directly-on-the-target-device) recommends that tensors be created directly on the target device by passing in `device="whatever"` instead of creating in the CPU and then moving to the target device.

File: src/training/sft.py [hunk 2]
```diff
@@ -158,7 +161,39 @@ def save_model(self) -> None:
...
             self.accelerator.save_state(path)
 
-    def step(self, batch: dict) -> None:
+    def warmup_step(self) -> None:
+        '''Preallocates memory for maximum sequence length in order to avoid nasty OOM shocks'''
+        model_config = self.model.gpt_neox.config
+        input_ids = torch.randint(
+            low=1,
+            high=model_config.vocab_size + 1,
+            # max_position_embeddings = 2048
```
Comment (0x000011b) [inline src/training/sft.py:170]: Do we need this for something?

**Review by 0x000011b — COMMENTED** [review]
Thanks, mostly LGTM! Left a few minor comments but nothing that blocks usage, will be testing before merging.

File: src/training/sft.py (hunk 1 above)
Comment (TearGosling) [reply src/training/sft.py:177]: Ah, right. Forgot about that. I can quickly fix that.

File: src/training/sft.py (hunk 2 above)
Comment (TearGosling) [reply src/training/sft.py:170]: The [PyTorch tips](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#pre-allocate-memory-in-case-of-variable-input-length) say that the fake batch that should be created preferably with randomly generated batches as inputs. `torch.randint()` in this case would make a tensor that resembles a "random" batch of tokenized text, in our case with the maximum sequence length. You could argue that I should do this for say, the starting and ending positions, but it might be more trouble than it's worth at that point.

Comment (0x000011b) [reply src/training/sft.py:170]: Yep, I meant the commented out `# max_position_embeddings = 2048`. Will we need that eventually? Or was it supposed to be un-commented/removed entirely?

File: src/training/sft.py [hunk 3]
```diff
@@ -158,7 +161,39 @@ def save_model(self) -> None:
...
+        rewards = torch.ones((self.args.batch_size, 1)).to("cuda")
+
+        outputs = sft_forward(
+            self.model,
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            start_positions=start_positions,
+            end_positions=end_positions,
+            rewards=rewards,
+            cringe_loss=self.args.use_cringe_loss
```
Comment (0x000011b) [inline src/training/sft.py:186]: These don't exist on `main`, so running this branch as-is doesn't work.

File: src/training/sft.py [hunk 4]
```diff
@@ -158,7 +161,39 @@ def save_model(self) -> None:
         else:
             self.accelerator.save_state(path)
 
-    def step(self, batch: dict) -> None:
+    def warmup_step(self) -> None:
+        '''Preallocates memory for maximum sequence length in order to avoid nasty OOM shocks'''
+        model_config = self.model.gpt_neox.config
```
Comment (0x000011b) [inline src/training/sft.py:166]: Needs to be:

```suggestion
        model_config = self.model.module.gpt_neox.config
```

Because of DeepSpeed.

File: src/training/sft.py [hunk 5]
```diff
@@ -374,6 +413,21 @@ def collate_fn(batches):
...
+        # These settings will need to be tuned.
+        # Maybe add these as a bunch of separate args later?
+        lora_config = LoraConfig(
+            task_type=TaskType.CAUSAL_LM,
+            inference_mode=False,
+            r=16,
+            lora_alpha=32,
+            lora_dropout=0.05,
+            merge_weights=False,
+        )
```
Comment (0x000011b) [inline src/training/sft.py:427]: Makes sense. For now, can we get just `r` converted into a CLI argument? I think it's the most important one we should be trying to tune - we can add in alpha and dropout later if relevant.

**Review by 0x000011b — CHANGES_REQUESTED** [review]
Got the chance to test the code and found some minor issues, but after fixing these it's good to merge.

**Review by 0x000011b — APPROVED** [review]
Awesome, LGTM now. Thanks.
